src/compiler/checker.ts

```diff
@@ -6487,6 +6487,28 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             symbolToNode: (symbol: Symbol, meaning: SymbolFlags, enclosingDeclaration?: Node, flags?: NodeBuilderFlags, tracker?: SymbolTracker) => withContext(enclosingDeclaration, flags, tracker, context => symbolToNode(symbol, context, meaning)),
         };
 
+        /**
+         * Unlike the utilities `setTextRange`, this checks if the `location` we're trying to set on `range` is within the
+         * same file as the active context. If not, the range is not applied. This prevents us from copying ranges across files,
+         * which will confuse the node printer (as it assumes all node ranges are within the current file).

```

Comment:
The printer’s single-file context has been a minor nuisance for code fixes and refactors in the past. For example, in an “add missing members from an incomplete `implements`” fix, we might write method signatures originating in two different files. How much more trouble would it be to fix the underlying assumption that every Node’s `originalNode` has the same source file?

Reply:
It's not even an issue with `.original`, it's an issue with assumptions about the text range - two of them. One is that, for literal strings and identifiers, unless some conditions are met, the printer just copies the input range to the output for them. AFAIK, this is to save on the escaping and unescaping work... I think. @rbuckton would have stronger opinions on that. The other is the token/node range bounds checking/assertions done in many services helpers. Specifically, we often assert that the last child of a node has the end position of its' parent - even inside the formatter. It's not clear to me that this is actually a good thing to assert on constructed node trees (though it is obviously true for parsed ones), but it is obviously not a good thing to assert on a tree with node ranges from a mixed set of input files, so we'd definitely need to revisit a lot of assertions in a bunch of services methods.